Panic: nil pointer deref in `ChannelsCollector` cache-refresh goroutine when `Logger` is uninitialized

**Affected versions:** v0.2.15 (latest tag); code path unchanged on `master`.

**Introduced in:** `0a2570a` ("collectors: tolerate DeadlineExceeded in all scrape-cycle
RPCs"), first released in v0.2.15.

## Summary

lndmon can crash on startup with a nil pointer dereference in the `ChannelsCollector`
cache-refresh goroutine. The package-global `Logger` is used by a goroutine that is
launched *before* `Logger` is initialized, so any early collector error dereferences a
nil logger.

Commit `0a2570a` changed several collector error paths from `errChan <- err` to a
log-then-check pattern (`Logger.Errorf(...)` / `Logger.Error(...)` followed by an
`IsDeadlineExceeded` guard). The intent was good — log transient `DeadlineExceeded`
errors and skip the scrape cycle instead of exiting — but it put a `Logger` use on code
that can run before `Logger` is assigned. The crash reported here is in
`ChannelsCollector`'s refresh goroutine, but the same nil-`Logger` exposure now exists in
every collector touched by that commit: `InfoCollector`, `ChannelsCollector`,
`WalletCollector`, `GraphCollector`, and `WtClientCollector`. Any of them erroring while
`Logger` is nil crashes the same way.

## Panic

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x58 pc=0x14a5944]

goroutine 30 [running]:
github.com/lightninglabs/lndmon/collectors.NewChannelsCollector.func1()
    /go/src/github.com/lightninglabs/lndmon/collectors/channels_collector.go:205 +0xe4
created by github.com/lightninglabs/lndmon/collectors.NewChannelsCollector in goroutine 1
    /go/src/github.com/lightninglabs/lndmon/collectors/channels_collector.go:198 +0x996
```

## Root cause — use-before-init ordering

1. `NewPrometheusExporter()` constructs the collectors. `NewChannelsCollector` *immediately*
   launches a cache-refresh goroutine (`collectors/channels_collector.go:198`) whose first
   loop iteration calls `refreshClosedChannelsCache()` → `ClosedChannels` with no initial
   delay.
2. The error branch of that loop logs unconditionally (`collectors/channels_collector.go:205`):
   ```go
   err := collector.refreshClosedChannelsCache()
   if err != nil {
       Logger.Errorf("ChannelsCollector refreshClosedChannelsCache failed with: %v", err)
       if !IsDeadlineExceeded(err) { errChan <- err }
   }
   ```
3. `Logger` is assigned only later, inside `PrometheusExporter.Start()` → `initLogRotator()`
   (`collectors/prometheus.go:154`), and only at the *end* of that function
   (`collectors/log.go:55`), after `os.MkdirAll(logDir, 0700)` (`collectors/log.go:33`)
   succeeds.

In `lndmon.go`'s `start()`, `NewPrometheusExporter(...)` (line 81) runs before
`exporter.Start()` (line 84) — so the collector goroutine is live while `Logger` is still
nil.

## Two ways it triggers

- **Race:** when lnd's RPC isn't fully ready yet, `ClosedChannels` errors in the window
  before `Start()` finishes logger init → `Logger.Errorf` on nil.
- **Permanent:** if `initLogRotator`'s `MkdirAll` fails (e.g. running under a read-only root
  filesystem with the default logdir `$HOME/.lndmon/logs` unwritable), `Start()` returns
  early and `Logger` is *never* assigned — so it stays nil for the whole process, and any
  collector error (startup, or a later transient `deadline exceeded`) panics.

## Reproduction

Run lndmon against an lnd node whose RPC isn't fully serving yet (or with an unwritable
`--prometheus.logdir`) so the first `ClosedChannels` scrape errors. The cache-refresh
goroutine panics instead of logging.

## Suggested fixes (any one resolves it)

1. Initialize logging (`initLogRotator`) *before* constructing collectors, so `Logger` is
   non-nil before any goroutine starts.
2. Don't launch the cache-refresh goroutine in `NewChannelsCollector` (construction); start
   it from `Start()` after logger init.
3. Treat `initLogRotator` failure as fatal in `Start()` *and* ensure no collector goroutine
   runs before it — i.e. don't leave `Logger` nil while goroutines are live.
4. Defensively nil-guard the error path (or inject a non-nil default/no-op logger), so a
   logging failure can't escalate a recoverable scrape error into a crash.

Fix #1 or #2 is the cleanest. The `Logger.Errorf` / `Logger.Error` calls added in
`0a2570a` (alongside the `IsDeadlineExceeded` handling) sit on paths that can run before
`Logger` exists. Whatever fix is chosen should cover all five collectors changed in that
commit, not just `ChannelsCollector`.

--- a/collectors/channels_collector.go
+++ b/collectors/channels_collector.go
@@ -194,7 +194,10 @@ func NewChannelsCollector(lnd lndclient.LightningClient, errChan chan<- error,
 		quit:                quitChan,
 	}
 
-	// Start a ticker to update the cache once per 10m
+	// Start a ticker to update the cache once per 10m. This is safe to
+	// launch here because NewPrometheusExporter initializes logging before
+	// constructing any collector, so the package-global Logger is ready by
+	// the time this goroutine can log on failure.
 	go func() {
 		ticker := time.NewTicker(cacheRefreshInterval)
 		defer ticker.Stop()

--- a/collectors/prometheus.go
+++ b/collectors/prometheus.go
@@ -94,7 +94,19 @@ func DefaultConfig() *PrometheusConfig {
 // the address to listen for Prometheus on and an lnd gRPC client.
 func NewPrometheusExporter(cfg *PrometheusConfig, lnd *lndclient.LndServices,
 	monitoringCfg *MonitoringConfig,
-	quitChan chan struct{}) *PrometheusExporter {
+	quitChan chan struct{}) (*PrometheusExporter, error) {
+
+	// Initialize logging before constructing any collectors. Some collectors
+	// launch background goroutines that log on failure, so the package-global
+	// Logger must be ready beforehand. A failure here is fatal: we return the
+	// error so lndmon exits instead of running with collectors that would
+	// dereference a nil Logger.
+	if err := initLogRotator(
+		filepath.Join(cfg.LogDir, defaultLogFilename),
+		defaultLogFileSize, defaultMaxLogFile,
+	); err != nil {
+		return nil, err
+	}
 
 	// We have six collectors and a htlc monitor running, so we buffer our
 	// error channel by 8 so that we do not need to consume all errors from
@@ -145,21 +157,12 @@ func NewPrometheusExporter(cfg *PrometheusConfig, lnd *lndclient.LndServices,
 		htlcMonitor:     htlcMonitor,
 		paymentsMonitor: paymentsMonitor,
 		errChan:         errChan,
-	}
+	}, nil
 }
 
 // Start registers all relevant metrics with the Prometheus library, then
 // launches the HTTP server that Prometheus will hit to scrape our metrics.
 func (p *PrometheusExporter) Start() error {
-	err := initLogRotator(
-		filepath.Join(p.cfg.LogDir, defaultLogFilename),
-		defaultLogFileSize,
-		defaultMaxLogFile,
-	)
-	if err != nil {
-		return err
-	}
-
 	Logger.Info("Starting Prometheus exporter...")
 	if p.lnd == nil {
 		return fmt.Errorf("cannot start PrometheusExporter without " +

--- a/lndmon.go
+++ b/lndmon.go
@@ -78,9 +78,12 @@ func start() error {
 
 	// Start our Prometheus exporter. This exporter spawns a goroutine
 	// that pulls metrics from our lnd client on a set interval.
-	exporter := collectors.NewPrometheusExporter(
+	exporter, err := collectors.NewPrometheusExporter(
 		cfg.Prometheus, &lnd.LndServices, &monitoringCfg, quit,
 	)
+	if err != nil {
+		return err
+	}
 	if err := exporter.Start(); err != nil {
 		return err
 	}

